File: backend/nutrition/dri.py
"""Dietary Reference Intake (DRI) table -- RDA/AI and UL by sex and age
bracket, for the micronutrients tracked by the USDA ingest pipeline.

Values are the current NASEM/National Academies and NIH Office of Dietary
Supplements Dietary Reference Intakes for adults, extended down to age 18
since that's this app's minimum onboarding age. Not applicable to pregnancy
or lactation, which use different DRI tables entirely.

Sources: NIH NCBI Bookshelf DRI reference reports (NBK208874, NBK114316,
NBK545442) and the National Academies' 2019 Sodium and Potassium report.
"""

from __future__ import annotations

from dataclasses import dataclass

from backend.nutrition.targets import Sex

AgeBracket = tuple[int, int]

# Boundaries line up with every documented DRI step-change for our nutrient
# set: magnesium steps up at 31, calcium/iron/B6 step at 51, vitamin D steps
# at 71.
AGE_BRACKETS: list[AgeBracket] = [(18, 30), (31, 50), (51, 70), (71, 200)]


def get_age_bracket(age_years: float) -> AgeBracket:
    for lo, hi in AGE_BRACKETS:
        if lo <= age_years <= hi:
            return (lo, hi)
    return AGE_BRACKETS[-1] if age_years > AGE_BRACKETS[-1][1] else AGE_BRACKETS[0]


@dataclass(frozen=True)
class NutrientTarget:
    rda: float
    unit: str
    ul: float | None = None  # Tolerable Upper Intake Level, where one exists


# Nutrients whose RDA/AI does not change across adult age brackets.
# Sodium and potassium AI are deliberately NOT stratified by adult age
# bracket -- the 2019 DRI revision explicitly declined to do so citing
# insufficient evidence, and sodium doesn't even split by sex.
_CONSTANT_RDA: dict[Sex, dict[str, tuple[float, str]]] = {
    "male": {
        "sodium_mg": (1500, "mg"),
        "potassium_mg": (3400, "mg"),
        "zinc_mg": (11, "mg"),
        "selenium_ug": (55, "ug"),
        "vitamin_a_rae_ug": (900, "ug"),
        "vitamin_c_mg": (90, "mg"),
        "vitamin_e_mg": (15, "mg"),
        "vitamin_k_ug": (120, "ug"),
        "thiamin_mg": (1.2, "mg"),
        "riboflavin_mg": (1.3, "mg"),
        "niacin_mg": (16, "mg"),
        "folate_dfe_ug": (400, "ug"),
        "vitamin_b12_ug": (2.4, "ug"),
        "choline_mg": (550, "mg"),
        "phosphorus_mg": (700, "mg"),
        "copper_mg": (0.9, "mg"),
        "manganese_mg": (2.3, "mg"),
    },
    "female": {
        "sodium_mg": (1500, "mg"),
        "potassium_mg": (2600, "mg"),
        "zinc_mg": (8, "mg"),
        "selenium_ug": (55, "ug"),
        "vitamin_a_rae_ug": (700, "ug"),
        "vitamin_c_mg": (75, "mg"),
        "vitamin_e_mg": (15, "mg"),
        "vitamin_k_ug": (90, "ug"),
        "thiamin_mg": (1.1, "mg"),
        "riboflavin_mg": (1.1, "mg"),
        "niacin_mg": (14, "mg"),
        "folate_dfe_ug": (400, "ug"),
        "vitamin_b12_ug": (2.4, "ug"),
        "choline_mg": (425, "mg"),
        "phosphorus_mg": (700, "mg"),
        "copper_mg": (0.9, "mg"),
        "manganese_mg": (1.8, "mg"),
    },
}

# Nutrients whose RDA changes by age bracket: calcium and iron step at 51
# (women only for both), magnesium steps at 31 (both sexes), vitamin D steps
# at 71 (both sexes), B6 steps at 51 (both sexes).
_BRACKET_RDA: dict[tuple[Sex, AgeBracket], dict[str, tuple[float, str]]] = {
    ("male", (18, 30)): {
        "calcium_mg": (1000, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (400, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.3, "mg"),
    },
    ("male", (31, 50)): {
        "calcium_mg": (1000, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (420, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.3, "mg"),
    },
    ("male", (51, 70)): {
        "calcium_mg": (1000, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (420, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.7, "mg"),
    },
    ("male", (71, 200)): {
        "calcium_mg": (1200, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (420, "mg"),
        "vitamin_d_ug": (20, "ug"),
        "vitamin_b6_mg": (1.7, "mg"),
    },
    ("female", (18, 30)): {
        "calcium_mg": (1000, "mg"),
        "iron_mg": (18, "mg"),
        "magnesium_mg": (310, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.3, "mg"),
    },
    ("female", (31, 50)): {
        "calcium_mg": (1000, "mg"),
        "iron_mg": (18, "mg"),
        "magnesium_mg": (320, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.3, "mg"),
    },
    ("female", (51, 70)): {
        "calcium_mg": (1200, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (320, "mg"),
        "vitamin_d_ug": (15, "ug"),
        "vitamin_b6_mg": (1.5, "mg"),
    },
    ("female", (71, 200)): {
        "calcium_mg": (1200, "mg"),
        "iron_mg": (8, "mg"),
        "magnesium_mg": (320, "mg"),
        "vitamin_d_ug": (20, "ug"),
        "vitamin_b6_mg": (1.5, "mg"),
    },
}

# Tolerable Upper Intake Levels. Constant across sex and adult age bracket
# for these eight nutrients per the NASEM DRI reports. Niacin and folate ULs
# apply only to synthetic/fortified intake (supplements, fortified food),
# not naturally occurring dietary niacin or food folate -- the UI warning
# for those two should be read with that caveat, not applied to total
# whole-food intake as strictly as the other six. Sodium has no formal UL;
# the 2019 Chronic Disease Risk Reduction (CDRR) value is used here as the
# upper-bound warning threshold instead.
UPPER_LIMITS: dict[str, float] = {
    "iron_mg": 45,
    "zinc_mg": 40,
    "selenium_ug": 400,
    "vitamin_a_rae_ug": 3000,
    "vitamin_d_ug": 100,
    "niacin_mg": 35,
    "folate_dfe_ug": 1000,
    "sodium_mg": 2300,
}


def get_micronutrient_targets(sex: Sex, age_years: float) -> dict[str, NutrientTarget]:
    bracket = get_age_bracket(age_years)
    combined: dict[str, tuple[float, str]] = {
        **_CONSTANT_RDA[sex],
        **_BRACKET_RDA[(sex, bracket)],
    }
    return {
        key: NutrientTarget(rda=value, unit=unit, ul=UPPER_LIMITS.get(key))
        for key, (value, unit) in combined.items()
    }

File: lib/nutrition/dri.ts
/**
 * Hardcoded US/Canada Dietary Reference Intake (DRI) table for adults,
 * keyed by (sex, age bracket). Values are RDAs/AIs from the NIH Office of
 * Dietary Supplements fact sheets. Only under-18 users are blocked upstream
 * in targets.ts, so no pediatric brackets are needed here.
 *
 * Nutrient keys match the keys used in public/data/foods.json's `nutrients`
 * object, so a food's per-100g values can be looked up against this table
 * with no key translation.
 *
 * Tolerable Upper Intake Levels (UL) are only populated for the nutrients
 * where exceeding them carries a meaningful, well-documented risk: iron,
 * zinc, selenium, vitamin A, vitamin D, niacin, folate, and sodium. Other
 * nutrients have `ul: null` — either no UL is established, or the risk of
 * over-supplementation from food alone is not clinically meaningful.
 *
 * Sodium has no true DRI "UL" — 2300 mg is the 2019 Chronic Disease Risk
 * Reduction (CDRR) intake, used here as the practical "don't exceed" value.
 */

export type Sex = "male" | "female";
export type AgeBracket = "19-30" | "31-50" | "51-70" | "71+";

export function getAgeBracket(age: number): AgeBracket {
  if (age < 31) return "19-30";
  if (age < 51) return "31-50";
  if (age < 71) return "51-70";
  return "71+";
}

export interface NutrientDri {
  /** Recommended Dietary Allowance, or Adequate Intake where no RDA is established. */
  rda: number;
  /** Tolerable Upper Intake Level, or null where none is tracked/established. */
  ul: number | null;
  unit: "mg" | "ug" | "mcgRAE" | "mcgDFE";
}

type BracketTable = Record<AgeBracket, NutrientDri>;
type DriTable = Record<string, Record<Sex, BracketTable>>;

const ALL_BRACKETS: AgeBracket[] = ["19-30", "31-50", "51-70", "71+"];

/** Same RDA/UL across every adult age bracket. */
function uniform(rda: number, ul: number | null, unit: NutrientDri["unit"]): BracketTable {
  return Object.fromEntries(ALL_BRACKETS.map((b) => [b, { rda, ul, unit }])) as BracketTable;
}

export const DRI_TABLE: DriTable = {
  sodiumMg: {
    male: uniform(1500, 2300, "mg"),
    female: uniform(1500, 2300, "mg"),
  },
  potassiumMg: {
    male: uniform(3400, null, "mg"),
    female: uniform(2600, null, "mg"),
  },
  calciumMg: {
    male: {
      "19-30": { rda: 1000, ul: null, unit: "mg" },
      "31-50": { rda: 1000, ul: null, unit: "mg" },
      "51-70": { rda: 1000, ul: null, unit: "mg" },
      "71+": { rda: 1200, ul: null, unit: "mg" },
    },
    female: {
      "19-30": { rda: 1000, ul: null, unit: "mg" },
      "31-50": { rda: 1000, ul: null, unit: "mg" },
      "51-70": { rda: 1200, ul: null, unit: "mg" },
      "71+": { rda: 1200, ul: null, unit: "mg" },
    },
  },
  ironMg: {
    male: uniform(8, 45, "mg"),
    female: {
      "19-30": { rda: 18, ul: 45, unit: "mg" },
      "31-50": { rda: 18, ul: 45, unit: "mg" },
      "51-70": { rda: 8, ul: 45, unit: "mg" },
      "71+": { rda: 8, ul: 45, unit: "mg" },
    },
  },
  magnesiumMg: {
    male: {
      "19-30": { rda: 400, ul: null, unit: "mg" },
      "31-50": { rda: 420, ul: null, unit: "mg" },
      "51-70": { rda: 420, ul: null, unit: "mg" },
      "71+": { rda: 420, ul: null, unit: "mg" },
    },
    female: {
      "19-30": { rda: 310, ul: null, unit: "mg" },
      "31-50": { rda: 320, ul: null, unit: "mg" },
      "51-70": { rda: 320, ul: null, unit: "mg" },
      "71+": { rda: 320, ul: null, unit: "mg" },
    },
  },
  zincMg: {
    male: uniform(11, 40, "mg"),
    female: uniform(8, 40, "mg"),
  },
  seleniumUg: {
    male: uniform(55, 400, "ug"),
    female: uniform(55, 400, "ug"),
  },
  phosphorusMg: {
    male: uniform(700, null, "mg"),
    female: uniform(700, null, "mg"),
  },
  copperMg: {
    male: uniform(0.9, null, "mg"),
    female: uniform(0.9, null, "mg"),
  },
  manganeseMg: {
    male: uniform(2.3, null, "mg"),
    female: uniform(1.8, null, "mg"),
  },
  vitAUg: {
    male: uniform(900, 3000, "mcgRAE"),
    female: uniform(700, 3000, "mcgRAE"),
  },
  vitCMg: {
    male: uniform(90, 2000, "mg"),
    female: uniform(75, 2000, "mg"),
  },
  vitDUg: {
    male: {
      "19-30": { rda: 15, ul: 100, unit: "ug" },
      "31-50": { rda: 15, ul: 100, unit: "ug" },
      "51-70": { rda: 15, ul: 100, unit: "ug" },
      "71+": { rda: 20, ul: 100, unit: "ug" },
    },
    female: {
      "19-30": { rda: 15, ul: 100, unit: "ug" },
      "31-50": { rda: 15, ul: 100, unit: "ug" },
      "51-70": { rda: 15, ul: 100, unit: "ug" },
      "71+": { rda: 20, ul: 100, unit: "ug" },
    },
  },
  vitEMg: {
    male: uniform(15, null, "mg"),
    female: uniform(15, null, "mg"),
  },
  vitKUg: {
    male: uniform(120, null, "ug"),
    female: uniform(90, null, "ug"),
  },
  thiaminMg: {
    male: uniform(1.2, null, "mg"),
    female: uniform(1.1, null, "mg"),
  },
  riboflavinMg: {
    male: uniform(1.3, null, "mg"),
    female: uniform(1.1, null, "mg"),
  },
  niacinMg: {
    male: uniform(16, 35, "mg"),
    female: uniform(14, 35, "mg"),
  },
  vitB6Mg: {
    male: {
      "19-30": { rda: 1.3, ul: null, unit: "mg" },
      "31-50": { rda: 1.3, ul: null, unit: "mg" },
      "51-70": { rda: 1.7, ul: null, unit: "mg" },
      "71+": { rda: 1.7, ul: null, unit: "mg" },
    },
    female: {
      "19-30": { rda: 1.3, ul: null, unit: "mg" },
      "31-50": { rda: 1.3, ul: null, unit: "mg" },
      "51-70": { rda: 1.5, ul: null, unit: "mg" },
      "71+": { rda: 1.5, ul: null, unit: "mg" },
    },
  },
  folateUg: {
    male: uniform(400, 1000, "mcgDFE"),
    female: uniform(400, 1000, "mcgDFE"),
  },
  vitB12Ug: {
    male: uniform(2.4, null, "ug"),
    female: uniform(2.4, null, "ug"),
  },
  cholineMg: {
    male: uniform(550, null, "mg"),
    female: uniform(425, null, "mg"),
  },
};

export function getDri(nutrientKey: string, sex: Sex, age: number): NutrientDri | undefined {
  return DRI_TABLE[nutrientKey]?.[sex]?.[getAgeBracket(age)];
}
